Disable killing the server by systemd

azat opened the pull request:
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Disable killing the server by systemd (that may lead to data loss when using Buffer tables)

### Documentation entry for user-facing changes
Default systemd's timeout for sending SIGKILL after SIGTERM is 1m30s
(TimeoutStopSec), which is can be not enough to wait for queries or
shutdown the storages.

And besides in this case shutdown_wait_unfinished server settings are
ignored.

So let's just disable this systemd logic and rely on
shutdown_wait_unfinished instead.

But note shutting down the storages can take a while, but it is better
to give it time instead of killing the process, since killing may lead
to data loss.

azat committed:
Disable killing the server by systemd (that may lead to data loss for Buffer)

Default systemd's timeout for sending SIGKILL after SIGTERM is 1m30s
(TimeoutStopSec), which is can be not enough to wait for queries or
shutdown the storages.

And besides in this case shutdown_wait_unfinished server settings are
ignored.

So let's just disable this systemd logic and rely on
shutdown_wait_unfinished instead.

But note shutting down the storages can take a while, but it is better
to give it time instead of killing the process, since killing may lead
to data loss.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Convert shutdown_wait_unfinished into ServerSettings

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Document shutdown_wait_unfinished/shutdown_wait_unfinished_queries in config.xml

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

robot-clickhouse-ci-1 commented:
<!-- automatic status comment for PR #54744 from azat/ClickHouse:server-shutdown -->
*This is an automated comment for commit 540d5a6371c24538165663ce95f70311b0dae68e with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/54744/540d5a6371c24538165663ce95f70311b0dae68e/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>✅ success</td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>⏳ pending</td></tr>
<tbody>
</table>

alexey-milovidov commented:
Ok.

PS. MergeTree tables, Atomic database engine, and even asynchronous inserts with default settings are safe for `kill -9`. If you kill the server at any time, there is no data loss or inconsistency. I will edit the description.

azat commented:
> PS. MergeTree tables, Atomic database engine, and even asynchronous inserts with default settings are safe for kill -9. If you kill the server at any time, there is no data loss or inconsistency.

Yes, I was talking about Buffer table first of all.

>I will edit the description.

Thanks

tavplubix commented:
> that may lead to data loss when using Buffer tables

That's expected behavior. It's totally fine to kill clickhouse-server with SIGKILL. Disabling it for systemd will not help, because there are other ways the server may get killed with signal 9 (like OOM)

azat commented:
> That's expected behavior. It's totally fine to kill clickhouse-server with SIGKILL.

Yes, but this is one of the most common "mistakes" that can lead to loosing data in Buffer table.

> Disabling it for systemd will not help, because there are other ways the server may get killed with signal 9 (like OOM)

Indeed, but OOM is another story, and usually if you have only clickhouse on the server it will not be triggered.
But in case of systemd with it's 1:30s timeout, it will also ignore `shutdown_wait_unfinished` setting.

This is all minor of course, but still.

azat committed:
Add ability to disable forwarding signals to the child by watchdog

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Disable forwarding signals by watchdog in systemd service

With default KillMode=control-group, systemd will send signals to all
processes in cgroup and this will lead to server will be terminated
forcefully due to double signal.

    2023.09.19 12:47:06.369090 [ 763 ] {} <Information> Application: Received termination signal (Terminated)
    2023.09.19 12:47:06.369141 [ 762 ] {} <Debug> Application: Received termination signal.
    2023.09.19 12:47:06.369215 [ 763 ] {} <Information> Application: Received termination signal (Terminated)
    2023.09.19 12:47:06.369225 [ 763 ] {} <Information> Application: This is the second termination signal. Immediately terminate.
    2023.09.19 12:47:06.400959 [ 761 ] {} <Information> Application: Child process exited normally with code 143.

Someone may naively think that, hey, I can change KillMode to
process/mixed, but this will not work either, because in this case
systemd cannot wait for the $MainPID (and main_pid_alien=true in
systemd's sources), because it is not a child of systemd, and this will
lead to double signal again:

    2023.09.19 16:24:19.694473 [ 3118 ] {} <Information> Application: Received termination signal (Terminated)
    2023.09.19 16:24:19.694894 [ 3118 ] {} <Information> Application: Received termination signal (Terminated)
    2023.09.19 16:24:19.695060 [ 3118 ] {} <Information> Application: This is the second termination signal. Immediately terminate.

And this is because it sends signal firstly on a normal termnation and
then when it cleans up left over processes:

    clickhouse-server.service: Found left-over process 3117 (clickhouse-serv) in control group while starting unit. Ignoring.

And yes, even though it prints "Ignoring" here (I guess it is related to
the fact that it can be ignored if the signal will not be handled)

Here is a proof of double signal by systemd:

    # pgrep clickhouse-serv | xargs strace  -e /kill -fp
    strace: Process 3117 attached with 469 threads
    [pid  3582] --- SIGTERM {si_signo=SIGTERM, si_code=SI_USER, si_pid=1, si_uid=0} ---
    [pid  3580] --- SIGCONT {si_signo=SIGCONT, si_code=SI_USER, si_pid=1, si_uid=0} ---
    [pid  3582] --- SIGCONT {si_signo=SIGCONT, si_code=SI_USER, si_pid=1, si_uid=0} ---
    [pid  3580] --- SIGTERM {si_signo=SIGTERM, si_code=SI_USER, si_pid=1, si_uid=0} ---
                                                                      ^^^
    [pid  3118] tgkill(3117, 3118, SIGTERM) = 0 # and this is a force termination

So yes, there is no other way except for disabling signal forwarding.

*Well, there is another way, but I guess it is will be unwelcome (even
though systemd can be configured in multiple ways right now, and there
is even systemd-oomd instead of clickhouse'es watchdog) - disable
watchdog completelly.*

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented on packages/clickhouse-server.service:
```diff
@@ -17,6 +17,13 @@ User=clickhouse
+TimeoutStopSec=infinity
+# Disable forwarding signals by watchdog, since with default systemd's
+# kill-mode control-group, systemd will send signal to all process in cgroup.
+Environment=CLICKHOUSE_WATCHDOG_NO_FORWARD=1
```
Actually this did not work with enabled watchdog, and this is the fix.
You will find descriptive commit message here about the reasoning - [`5b40c81` (#54744)](https://github.com/ClickHouse/ClickHouse/pull/54744/commits/5b40c81ac6bc7c545d942963550ed2347c267b58)

azat commented:
>Integration tests (tsan) [5/6] — fail: 1, passed: 546

- `test_backup_restore_on_cluster/test.py::test_stop_other_host_during_backup` - #54816
